### PR TITLE
Implemented numeric-separator

### PR DESCRIPTION
**Pull request by kisbg**
Releated test262 test-cases added

JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 1]
```diff
@@ -293,7 +293,8 @@ static const uint8_t ecma_uint4_clz[] = { 4, 3, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 0,
 static ecma_number_t
 ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 string */
                                      const lit_utf8_byte_t *end_p, /**< end of utf-8 string  */
-                                     uint32_t radix) /**< radix */
+                                     uint32_t radix, /**< radix */
+                                     uint8_t has_underscore)  /**< string can contains "_" character */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:297]: You should use a flag or `bool` type.

File: jerry-core/parser/js/js-lexer.h [hunk 2]
```diff
@@ -270,6 +270,12 @@ typedef enum
   LEXER_STRING_RAW = (1u << 1),           /**< raw string ECMAScript v6, 11.8.6.1: TVR */
 } lexer_string_options_t;
 
+typedef enum
+{
+  LEXER_LIT_LOCATION_NO_OPTS = 0,
+  LEXER_LIT_LOCATION_HAS_ESCAPE = (1u << 0),
+  LEXER_LIT_LOCATION_HAS_UNDERSCORE = (1u << 1),
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.h:277]: Has escape and has underscore seems mutually exclusive. Btw, why do we exactly need this flag?

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/base/ecma-helpers-number.c [hunk 3]
```diff
@@ -1063,7 +1063,7 @@ ecma_number_parse_float (const lit_utf8_byte_t *string_buff, /**< routine's firs
   }
 
   /* 5. */
-  ecma_number_t ret_num = ecma_utf8_string_to_number (start_p, (lit_utf8_size_t) (end_p - start_p));
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-number.c:1066]: all `0` argumentum should be `false`

File: jerry-core/parser/js/js-lexer.c [hunk 4]
```diff
@@ -661,7 +661,7 @@ lexer_parse_identifier (parser_context_t *context_p, /**< context */
   parser_line_counter_t column = context_p->column;
   const uint8_t *source_end_p = context_p->source_end_p;
   size_t length = 0;
-  uint8_t has_escape = false;
+  uint8_t has_escape = LEXER_LIT_LOCATION_NO_OPTS;
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:664]: Please revert all the `has_escape changes`

File: jerry-core/parser/js/js-lexer.h [hunk 5]
```diff
@@ -292,7 +298,7 @@ typedef struct
   const uint8_t *char_p;                     /**< start of identifier or string token */
   prop_length_t length;                      /**< length or index of a literal */
   uint8_t type;                              /**< type of the current literal */
-  uint8_t has_escape;                        /**< has escape sequences */
+  uint8_t flags;                             /**< has flags */
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.h:301]: ditto.

File: jerry-core/parser/js/js-lexer.c [hunk 6]
```diff
@@ -2543,15 +2661,15 @@ lexer_construct_number_object (parser_context_t *context_p, /**< context */
...
 #if ENABLED (JERRY_BUILTIN_BIGINT)
   if (JERRY_LIKELY (context_p->token.extra_value != LEXER_NUMBER_BIGINT))
   {
 #endif /* ENABLED (JERRY_BUILTIN_BIGINT) */
     if (context_p->token.extra_value < LEXER_NUMBER_OCTAL)
     {
       num = ecma_utf8_string_to_number (context_p->token.lit_location.char_p,
-                                        length);
+                                        length,
+                                        context_p->token.lit_location.flags);
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:2672]: This should be always true.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/parser/js/js-lexer.c [hunk 7]
```diff
@@ -1280,15 +1280,37 @@ lexer_parse_string (parser_context_t *context_p, /**< context */
  */
 static inline void
 lexer_parse_octal_number (parser_context_t *context_p, /** context */
-                          const uint8_t **source_p) /**< current source position */
+                          const uint8_t **source_p, /**< current source position */
+                          bool is_legary) /**< is legacy octal number */
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:1284]: typo.

File: jerry-core/parser/js/js-lexer.c [hunk 8]
```diff
@@ -3472,7 +3580,7 @@ lexer_compare_identifiers (parser_context_t *context_p, /**< context */
     return lexer_compare_identifier_to_chars (right_p->char_p, left_p->char_p, length);
   }
 
-  if (!right_p->has_escape)
+  if (!left_p->has_escape)
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:3583]: Is this change intentional?

File: jerry-core/parser/js/js-lexer.h [hunk 9]
```diff
@@ -292,7 +292,7 @@ typedef struct
   const uint8_t *char_p;                     /**< start of identifier or string token */
   prop_length_t length;                      /**< length or index of a literal */
   uint8_t type;                              /**< type of the current literal */
-  uint8_t has_escape;                        /**< has escape sequences */
+  uint8_t has_escape;                             /**< has escape sequences */
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.h:295]: PLease revert this whitespace change.

File: jerry-core/parser/js/js-scanner-util.c [hunk 10]
```diff
@@ -1983,7 +1983,6 @@ scanner_check_variables (parser_context_t *context_p) /**< context */
     literal.length = data_p[1];
     literal.type = LEXER_IDENT_LITERAL;
     literal.has_escape = (data_p[0] & SCANNER_STREAM_HAS_ESCAPE) ? 1 : 0;
-
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-util.c:1986]: Please do not remove this line.

File: jerry-core/parser/js/js-lexer.c [hunk 11]
```diff
@@ -1280,15 +1280,37 @@ lexer_parse_string (parser_context_t *context_p, /**< context */
...
+      if (is_legacy ||
+          *source_p  == context_p->source_end_p ||
+          *source_p[0] < LIT_CHAR_0 ||
+          *source_p[0] > LIT_CHAR_7)
+      {
+        parser_raise_error (context_p, PARSER_ERR_INVALID_UNDERSCORE_USE_IN_NUMBER);
+      }
+      continue;
+    }
+#endif /* ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:1311]: This function is almost identical to `lexer_check_underscore`. Cannot we combine it?

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 12]
```diff
@@ -293,7 +293,9 @@ static const uint8_t ecma_uint4_clz[] = { 4, 3, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 0,
 static ecma_number_t
 ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 string */
                                      const lit_utf8_byte_t *end_p, /**< end of utf-8 string  */
-                                     uint32_t radix) /**< radix */
+                                     uint32_t radix, /**< radix */
+                                     bool allow_underscore)  /**< true - if '_' characters are allowed in the string
+                                                                  false - otherwise */
 {
   JERRY_ASSERT (radix == 2 || radix == 8 || radix == 16);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:300]: Since radix is very limited, maybe we could use a flag bit to represent `allow_underscore`.

File: jerry-core/ecma/base/ecma-helpers.h [hunk 13]
```diff
@@ -550,7 +550,9 @@ ecma_native_pointer_t *ecma_get_native_pointer_value (ecma_object_t *obj_p, void
 bool ecma_delete_native_pointer_property (ecma_object_t *obj_p, void *info_p);
 
 /* ecma-helpers-conversion.c */
-ecma_number_t ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, lit_utf8_size_t str_size);
+ecma_number_t ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p,
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers.h:553]: You can keep the first two arguments in the same line.

File: jerry-core/ecma/operations/ecma-bigint.c [hunk 14]
```diff
@@ -73,6 +73,7 @@ ecma_bigint_parse_string (const lit_utf8_byte_t *string_p, /**< string represena
                           lit_utf8_size_t size, /**< string size */
                           uint32_t options) /**< ecma_bigint_parse_string_options_t option bits */
 {
+  bool allow_underscore = options & ECMA_BIGINT_PARSE_ALLOW_UNDERSCORE;
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-bigint.c:76]: Move this after `uint32_t sign = ...` line.

File: jerry-core/parser/js/js-lexer.c [hunk 15]
```diff
@@ -1276,26 +1276,43 @@ lexer_parse_string (parser_context_t *context_p, /**< context */
...
+           && *source_p[0] <= largest_number)
+    {
+      *source_p += 1;
+    }
+#if ENABLED (JERRY_ESNEXT)
+    if (*source_p != source_end_p && *source_p[0] == LIT_CHAR_UNDERSCORE)
+    {
+      *source_p += 1;
+      if (is_legacy || *source_p[0] == LIT_CHAR_UNDERSCORE ||
+          *source_p[0] > largest_number || *source_p[0] < LIT_CHAR_0)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1305]: Invalid syntax.

File: jerry-core/parser/js/js-lexer.c [hunk 16]
```diff
@@ -1482,7 +1514,6 @@ lexer_parse_number (parser_context_t *context_p) /**< context */
   context_p->token.lit_location.length = (prop_length_t) length;
   PARSER_PLUS_EQUAL_LC (context_p->column, length);
   context_p->source_p = source_p;
-
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1485]: Don't remove this line.

File: jerry-core/parser/js/js-lexer.c [hunk 17]
```diff
@@ -2543,15 +2574,13 @@ lexer_construct_number_object (parser_context_t *context_p, /**< context */
   ecma_number_t num;
   uint32_t literal_index = 0;
   prop_length_t length = context_p->token.lit_location.length;
-
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:2546]: Ditto

**Review by zherczeg — CHANGES_REQUESTED** [review]
Please add an esnext test, especially for cases with syntax error.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 18]
```diff
@@ -294,8 +294,12 @@ static ecma_number_t
 ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 string */
                                      const lit_utf8_byte_t *end_p, /**< end of utf-8 string  */
                                      uint32_t radix) /**< radix */
+
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:297]: No need this newline.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 19]
```diff
@@ -294,8 +294,12 @@ static ecma_number_t
...
 {
+  bool allow_underscore = (radix % 2) ? radix-- : false;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:299]: I find this strange, while it is true that it works in this case. I would prefer:

```
bool allow_underscore = (radix & 0x1);
radix &= (uint32_t) ~0x1;
```
Maybe we could add a define constant for 0x1

Furthermore guard this with JERRY_ESNEXT which makes the `JERRY_UNUSED (allow_underscore)` below unnecessary.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 20]
```diff
@@ -364,7 +376,9 @@ ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 st
  */
 ecma_number_t
 ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
-                            lit_utf8_size_t str_size) /**< string size */
+                            lit_utf8_size_t str_size, /**< string size */
+                            bool allow_underscore) /**< true - if '_' characters are allowed in the string
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:380]: This should be an uint32_t parameter which contains the defined flag.
Then you can:
`return ecma_utf8_string_to_number_by_radix (str_p + 2, end_p, 16 | underscore_flag);`
If you really want to keep the bool, you can create a local variable which contains the flag depending on the bool.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 21]
```diff
@@ -295,7 +300,12 @@ ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 st
                                      const lit_utf8_byte_t *end_p, /**< end of utf-8 string  */
                                      uint32_t radix) /**< radix */
 {
+#if ENABLED (JERRY_ESNEXT)
+  bool allow_underscore = (radix & ALLOW_UNDERSCORE);
+#endif /* ENABLED (JERRY_ESNEXT) */
+  radix &= (uint32_t) ~ALLOW_UNDERSCORE;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:306]: You can move this line up into the `#if`

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 22]
```diff
@@ -280,6 +280,11 @@ static const uint8_t ecma_uint4_clz[] = { 4, 3, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 0,
  */
 #define EPSILON 0.0000001
 
+/**
+ * it's a constant that check ecma_utf8_string_to_number_by_radix radix arguments
+ */
+#define ALLOW_UNDERSCORE 0x1
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:286]: Add a  JERRY_ESNEXT guard around it.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 23]
```diff
@@ -364,7 +380,9 @@ ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 st
  */
 ecma_number_t
 ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
-                            lit_utf8_size_t str_size) /**< string size */
+                            lit_utf8_size_t str_size, /**< string size */
+                            uint32_t allow_underscore_option) /**< true - if '_' characters are allowed in the string
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:384]: The comment is invalid. Also the name should simply be `options` since we may add more options in the future.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 24]
```diff
@@ -280,6 +280,11 @@ static const uint8_t ecma_uint4_clz[] = { 4, 3, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 0,
  */
 #define EPSILON 0.0000001
 
+/**
+ * it's a constant that check ecma_utf8_string_to_number_by_radix radix arguments
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:284]: Comment: `Underscore is ignored when this option is passed.`

I know the names above does not follow the syntax (they are very old), but please rename
ALLOW_UNDERSCORE -> ECMA_CONVERSION_ALLOW_UNDERSCORE

File: jerry-core/ecma/base/ecma-helpers.h [hunk 25]
```diff
@@ -96,6 +96,15 @@ typedef enum
   ECMA_STRING_FLAG_MUST_BE_FREED = (1 << 3), /**< The returned buffer must be freed */
 } ecma_string_flag_t;
 
+/**
+ * Status flags for allowing underscore in number create
+ */
+typedef enum
+{
+  PREFIX_NO_UNDERSCORE = 0, /**< underscore character is not allowed */
+  PREFIX_ALLOW_UNDERSCORE = (1 << 0), /**< underscore character is allowed */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers.h:105]: No need this flag if we have `ECMA_CONVERSION_ALLOW_UNDERSCORE`. Maybe just move the definition here:

```
typedef enum
{
  ECMA_CONVERSION_NO_OPTIONS = 0, /**< no options */
  ECMA_CONVERSION_IGNORE_UNDERSCORE = (1 << 0), /**< ignore underscore characters */
}
```
And call it ignore underscore.

File: jerry-core/parser/js/js-parser-util.c [hunk 26]
```diff
@@ -937,6 +937,10 @@ parser_error_to_string (parser_error_t error) /**< error code */
     {
       return "Missing exponent part.";
     }
+    case PARSER_ERR_INVALID_UNDERSCORE_USE_IN_NUMBER:
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-util.c:940]: PARSER_ERR_INVALID_UNDERSCORE_IN_NUMBER

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 27]
```diff
@@ -364,7 +375,9 @@ ecma_utf8_string_to_number_by_radix (const lit_utf8_byte_t *str_p, /**< utf-8 st
  */
 ecma_number_t
 ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
-                            lit_utf8_size_t str_size) /**< string size */
+                            lit_utf8_size_t str_size, /**< string size */
+                            uint32_t options) /**< true - if '_' characters are allowed in the string
+                                                                   false - otherwise */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:380]: Comment is invalid.

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 28]
```diff
@@ -550,6 +575,13 @@ ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
       {
         digit_value = (*str_p - LIT_CHAR_0);
       }
+#if ENABLED (JERRY_ESNEXT)
+      else if (*str_p == LIT_CHAR_UNDERSCORE && options)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:579]: `(options & ECMA_CONVERSION_ALLOW_UNDERSCORE)`
Options is a flag set not a boolean.

File: jerry-core/ecma/base/ecma-helpers-number.c [hunk 29]
```diff
@@ -1059,7 +1059,9 @@ ecma_number_parse_float (const lit_utf8_byte_t *string_buff, /**< routine's firs
...
+  ecma_number_t ret_num = ecma_utf8_string_to_number (start_p,
+                                                      (lit_utf8_size_t) (end_p - start_p),
+                                                      0);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-number.c:1064]: If this is not more than 120 chars, I would keep the original single line form.

File: jerry-core/ecma/builtin-objects/ecma-builtin-date.c [hunk 30]
```diff
@@ -68,7 +68,9 @@ ecma_date_parse_date_chars (const lit_utf8_byte_t **str_p, /**< pointer to the c
     }
   }
 
-  ecma_number_t parsed_number = ecma_utf8_string_to_number (str_start_p, (lit_utf8_size_t) (*str_p - str_start_p));
+  ecma_number_t parsed_number = ecma_utf8_string_to_number (str_start_p,
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-date.c:71]: Ditto, keep 1 line (there are other cases like this, please fix them all)

File: jerry-core/ecma/operations/ecma-bigint.c [hunk 31]
```diff
@@ -120,7 +121,7 @@ ecma_bigint_parse_string (const lit_utf8_byte_t *string_p, /**< string represena
     return ECMA_BIGINT_ZERO;
   }
 
-  while (string_p < string_end_p && *string_p == LIT_CHAR_0)
+  while (string_p < string_end_p && (*string_p == LIT_CHAR_0 || (*string_p == LIT_CHAR_UNDERSCORE && allow_underscore)))
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-bigint.c:124]: Could give me an example when the `(*string_p == LIT_CHAR_UNDERSCORE && allow_underscore)` part is needed?

File: jerry-core/parser/js/js-lexer.c [hunk 32]
```diff
@@ -1298,26 +1298,43 @@ lexer_parse_string (parser_context_t *context_p, /**< context */
...
+ * Check number
  */
-static inline void
-lexer_parse_octal_number (parser_context_t *context_p, /** context */
-                          const uint8_t **source_p) /**< current source position */
+static void
+lexer_check_numbers (parser_context_t *context_p, /**< context */
+                     const uint8_t **source_p, /**< source_pointer */
+                     const uint8_t *source_end_p, /**< end of the source */
+                     const ecma_char_t largest_number, /**< largest valid number */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-lexer.c:1307]: comment: largest allowed number
Btw, I would prefer "digit_max" since it represents the maximum digit, rather than the result number

File: jerry-core/parser/js/js-parser-util.c [hunk 33]
```diff
@@ -937,6 +937,10 @@ parser_error_to_string (parser_error_t error) /**< error code */
     {
       return "Missing exponent part.";
     }
+    case PARSER_ERR_INVALID_UNDERSCORE_IN_NUMBER:
+    {
+      return "Invalid use of underscore in number";
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-util.c:942]: `Invalid use of underscore character in number literals`

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/operations/ecma-bigint.c (hunk 31 above)
Comment (kisbg) [reply jerry-core/ecma/operations/ecma-bigint.c:124]: without the new expression this input will cause a crash : 0x0_0n

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 34]
```diff
@@ -491,6 +510,11 @@ ecma_utf8_string_to_number (const lit_utf8_byte_t *str_p, /**< utf-8 string */
         digit_seen = true;
         digit_value = (*str_p - LIT_CHAR_0);
       }
+      else if (*str_p == LIT_CHAR_UNDERSCORE && options)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c:513]: `(options & ECMA_CONVERSION_ALLOW_UNDERSCORE)` similar to above

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c [hunk 35]
```diff
@@ -287,7 +287,9 @@ ecma_builtin_json_parse_number (ecma_json_token_t *token_p) /**< token argument
   }
 
   token_p->type = TOKEN_NUMBER;
-  token_p->u.number = ecma_utf8_string_to_number (start_p, (lit_utf8_size_t) (current_p - start_p));
+  token_p->u.number = ecma_utf8_string_to_number (start_p,
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.c:290]: One line should be enough.

**Review by zherczeg — CHANGES_REQUESTED** [review]
Only minor changes needed.

**Review by zherczeg — APPROVED** [review]
LGTM

**Review by rerobika — APPROVED** [review]
LGTM
